resolver_wrapper.go

```diff
@@ -23,127 +23,98 @@ import (
 	"strings"
 	"sync"
 
-	"google.golang.org/grpc/balancer"
 	"google.golang.org/grpc/internal/channelz"
 	"google.golang.org/grpc/internal/grpcsync"
 	"google.golang.org/grpc/internal/pretty"
 	"google.golang.org/grpc/resolver"
 	"google.golang.org/grpc/serviceconfig"
 )
 
-// resolverStateUpdater wraps the single method used by ccResolverWrapper to
-// report a state update from the actual resolver implementation.
-type resolverStateUpdater interface {
-	updateResolverState(s resolver.State, err error) error
-}
```

Comment:
Do you have suggestions for how else to handle this?  I think these things are very closely tied together, and without this, we can have deadlocks due to lock ordering.

Reply:
I tried spending a little bit of time to make the resolver wrapper handle enter and exit idle mode. The wrapper in this change also handles creation of the new resolver and skipping updates from the old resolver. It needs another wrapper, I called it `genWrapper` to indicate that it wraps one generation of the underlying resolver. The changes are not very pretty, but I guess, with a little more time, it can be made to look a little better. But it does seem to work:
https://github.com/grpc/grpc-go/compare/master...easwars:resolver_wrapper_handle_enter_exit_idle?expand=1

Taking the approach of making all the wrappers handle enter and exit idle modes does make the wrappers a bit more verbose, but I feel it might be less complicated in the sense that the wrappers won't be reaching into and grabbing `cc.mu`, and won't have any imposed lock ordering that spans types. Let me know what you think.

Reply:
I have some concerns about your diffs:

- I am pretty sure the same deadlock is still there:

1. Resolver calls `UpdateState()`, grabs `ccr.mu`, enqueues work into serializer.
1. Channel does `EnterIdleMode` and takes `cc.mu`.
1. `UpdateState`'s work item gives up `ccr.mu`.
1. Channel calls `ccr.enterIdleMode` which grabs `ccr.mu` and then blocks on `serializer.Done()`.
1. Serializer calls `UpdateState`, which needs `cc.mu`.

- The `genWrapper` still reaches into its parent and grabs its mutex, which is potentially just as problematic readability/correctness-wise as the `ccr` reaching into its parent for `cc.mu`.

- How are you synchronizing calls _into_ the resolver?  It seems like `resolver.ResolveNow` and `resolver.Close` could happen concurrently or `ResolveNow`-after`Close`, which is not desired / illegal.

- The use of the serializer is odd.  You always hold a lock and release it once you're in the serializer, meaning only one thing can be in it at a time.  Why not just remove it and get the lock?  That would remove the deadlock that happens above, but.....

- I think you can also end up with the resolver able to call `cc.updateResolverState` _while_ entering idle mode, and without other changes, could cause problems.

This change went through a lot of iteration and has a number of improvements vs. master, and while I'm fine with continuing discussions about other ways of doing things, this fixes a deadlock that is blocking our release, so I would appreciate it if we can push forward with this approach (unless you spot any non-theoretical problems with it), and consider refining or revising it as follow-ups.

Reply:
My diff was not put together with the rigor required when actually sending for review. So, apologies for the many things that I missed (actually I didn't realize I missed so many things).

> so I would appreciate it if we can push forward with this approach (unless you spot any non-theoretical problems with it), and consider refining or revising it as follow-ups

The main problem I have is that now `cc.mu` is held when handling every update from the balancer and resolver. It is also held every time `cc` has to call into the resolver or the balancer through the appropriate wrapper. This makes me very nervous. Although I cannot point out any actual possible races or deadlocks (and clearly all our tests are passing), I feel that maybe we could have gone with the minimum change required to fix the deadlock and made the other changes subsequently. Also, not being around when the many iterations happened clearly doesn't help me.

Reply:
> The main problem I have is that now cc.mu is held when handling every update from the balancer and resolver

This is a good thing.  We block the channel from doing anything big and scary (e.g. entering idle mode) while the plugin is updating our state.

> It is also held every time cc has to call into the resolver or the balancer through the appropriate wrapper

I don't think this is true.  They need to grab it to read the wrapper, since the wrappers can come and go while entering/exiting idle, but these are all non-RPC events and grabbing the mutex during these events should be reasonable.

> Although I cannot point out any actual possible races or deadlocks (and clearly all our tests are passing), I feel that maybe we could have gone with the minimum change required to fix the deadlock

I'm not sure what that minimum change would be.  Like I said, this change took a _lot_ of iterations, and I definitely tried to start small, but kept running into other deadlocks or races.  I think the sync model we have now has some fundamental flaws that need structural changes to address.  It's not a simple problem, either, because we have a synchronous path from the channel to the resolver, back to the channel, to the LB policy, and back to the resolver through the channel, and during the call to the LB policy, we need to be able to handle other calls coming from it.